### ISSUE TITLE
Text Input Widget

### ISSUE THREAD

**Issue by dbrgn** [fdehau/tui-rs#34]
Just found your library and it looks awesome!

I'm aware that you wrote "the library does not provide any input handling nor any event system", but is there any example of a text input widget?

I'm trying to write a very simple chat application, so I'd just need a text input at the bottom and a scrolling text view at the top.

**Comment by fdehau** [comment]
I've added a very simple example of input handling.  The description is at the top of the file. I don't want to go in too much details regarding the input handling since the library does not make any assumptions about how you manage the state in your application nor how you want to get the input of the user.

As a simple warning, the examples do not illustrate necessarily some best practices but provide the minimum components to start drawing something on your terminal. I'm sure that there are clever way to do what is shown.

However, if you find yourself blocked trying to achieve something with the library, i would love to get some feedback (issues like this one or PR are welcomed)

**Comment by makoConstruct** [comment]
@fdehau yeah, I do request an example that shows best practices. I'm having a lot of friction figuring out how to respond to sigint and then close the app. I wouldn't know where to look to get ctrl+z to work properly, and it looks like, without making an RAII thing to deal with it I'm pretty sure my apps will permanently hide the user's cursor whenever they terminate with a panic (the examples call `show_cursor` explicitly at the end of main, but if the program panics before then, that will be skipped)

We need more guidance. I think it would make sense to put some of that guidance in the text_input example.

Side-question: I can't figure out how you're making stdin unblocks every time the user types a character. When I handle stdin directly it only unblocks when the user types a newline. It would be nice to know how you do that, whatever it is, I didn't spot it when I looked at the source.

**Comment by fdehau** [comment]
Hi,

I haven't experimented with signal handling in the context of this crate yet but a quick search yielded [this](https://github.com/BurntSushi/chan-signal) which links to `signal-hook` as a possible solution. I'm in the process of cleaning the examples so I might give it a shot.

Regarding the issue of the panic, I'm not sure how an RAII pattern would help. To my knowledge, you might be able to implement custom behavior before the panic runtime is invoked using [set_hook](https://doc.rust-lang.org/std/panic/fn.set_hook.html), but I haven't tried myself either. I understand that having the cursor disappear is annoying for the end user but at the same time panics should happen rarely. 

Finally the behavior of stdin is modified by `termion` in the example and I would encourage you to read the `termion` documentation and its README if you haven't already. If you handle `stdin` yourself, I think that you get the default `readline` behavior which operates on lines.

Input handling is very specific to the backend you use and is beyond the scope of this crate. As stated in the README and my previous comment, I've only focused my efforts on providing a common rendering interface.

Nevertheless, if that answer is not enough to solve your issues I'm happy to provide more guidance if I'm able to.

**Comment by makoConstruct** [comment]
>Regarding the issue of the panic, I'm not sure how an RAII pattern would help

something like

```
struct CursorRevealGuard();
impl Drop for CursorRevealGuard{ fn drop(){ drop(terminal.show_cursor()); } }
fn hide_cursor(t:&mut Terminal)-> Result<CursorRevealGuard,   > { terminal.hide_cursor() }

fn main() {
    let mut terminal = ...;
    let guard = hide_cursor(&mut terminal);
}
```

If a panic occurs in main's thread, the guard will be `drop`ped during the unwinding and the cursor will be set to show again. We drop the result of showing instead of unwrapping it, because throwing a panic from drops does nothing good (it would abort)

Thanks. I'll take a closer look at what termion's doing.

**Comment by fdehau** [comment]
Ah ok, I've had ignored this pattern because I thought I should not trigger any `panic!` when `drop` is called on `Terminal`. But I can simply ignore the error and attempt to show the cursor again in a best effort strategy. Thanks for the suggestion. It is also nicer since you are not required to call `show_cursor` anymore at the end of your program. I've implemented this in #74.
